Feature/#186 only relevant stack trace paths

KGonzalezASC opened the pull request:
To avoid overhauling Step.cs , I made changes to  ExceptionExtensions.cs to filter the needed output.

Here are the changes:

Commit 1:
- renamed namespace for ExceptionExtensions.cs as suggested by Rider
- renamed exceptionPathRegex to KoanExceptionPathRegex better clarify context
- Updated Regex object to pattern match based on the Koan naming scheme starting with “About”
- Output should now only show the relevant Koans Excercise

Commit 2:
- Added a separate “GetStackTracePath” like method that only provides an output based on a pattern match for the “Asserts” (which appears to be Xunits naming scheme.) It has to include the About negative look-behind to prevent (notably AboutAsserts.cs) a Koan from returning in the match
- Updated console to output in format @kilasuit suggested.


Please let me know if there are any issues.

KGonzalezASC committed:
- renamed namespace for Extension classes as suggested by Rider
- renamed exceptionPathRegex to KoanExceptionPathRegex better clarify context
- Updated Regex object to pattern match based on the Koan naming scheme starting with “About”
-Output should now only show the relevant Koans Excercise

KGonzalezASC committed:
- Added Seperate “GetStackTracePath” like method that only provides an output based on a  pattern match for the “Asserts” (which appears to be Xunits naming scheme) it has to include the About negative lookbehind to prevent (notably AboutAssert.cs) a Koan from returning in the match
- Updated console to output in format @kilasuit suggested.

KGonzalezASC committed:
- Added Seperate “GetStackTracePath” like method that only provides an output based on a  pattern match for the “Asserts” (which appears to be Xunits naming scheme) it has to include the About negative lookbehind to prevent (notably AboutAssert.cs) a Koan from returning in the match
- Updated console to output in format @kilasuit suggested.

MaxAtoms commented on Engine/Extensions/ExceptionExtensions.cs:
```diff
@@ -1,15 +1,20 @@
-        private static readonly Regex exceptionPathRegex = new Regex(@"(\w*\.cs:line \d*)");
-
+        private static readonly Regex KoanExceptionPathRegex = new(@"(About\w*\.cs:line \d*)");
+        private static readonly Regex AssertExceptionPathRegex = new(@"\w*Asserts\.cs:line \d*(?<!About\w*\.cs:line \d*)");
```
Huh, I never used a negative look-behind myself. Learned something today :)
We certainly can leave it like that if you want to. I just think that it is kind of unnecessary to parse the string twice and to sort of implicitly exclude the Koans here.

We could parse the string once via Regex and then separate Koan paths from the other ones.
Something like this:

```csharp
public record StackTracePaths(IEnumerable<string> KoanPaths, IEnumerable<string> OtherPaths);

private static readonly Regex ExceptionPathRegex = new(@"\w+\.cs:line \d*");
        
public static StackTracePaths GetStackTracePaths(this Exception exception)
{
	var allPaths = ExceptionPathRegex
		.Matches(exception.StackTrace ?? "").Select(x => x.Value)
		.ToList();
	
	var koanPaths = allPaths.Where(x => x.StartsWith("About")).ToList();

	return new StackTracePaths( koanPaths, allPaths.Except(koanPaths) );
}
```

KGonzalezASC replied:
I agree that this approach would work well! I should have thought to use Records, but my college coursework is always behind on newer .NET features, so it didn't come to mind 😅

I will have an updated pull request made later tonight with these changes. I also tweaked the linq expression to exclude the list allocations.

MaxAtoms replied:
While it is not necessary to use a record, it is nice to have the positional syntax and the `Deconstruct` method here. We could use a tuple, but I do not like to pass tuples outside of classes. They also tend to clutter your method definitions.

> but my college coursework is always behind on newer .NET features

I do not blame them. In the last years the language is gaining features fast 😅

KGonzalezASC commented on Engine/Extensions/ExceptionExtensions.cs:
```diff
@@ -1,15 +1,20 @@
     {
-        private static readonly Regex exceptionPathRegex = new Regex(@"(\w*\.cs:line \d*)");
-
+        private static readonly Regex KoanExceptionPathRegex = new(@"(About\w*\.cs:line \d*)");
```
The capture group helps to immediately identify the matching substring. Given that the substrings are small, they can be excluded. 

I will change the quantifier, my apologies.

MaxAtoms replied:
No need to apologize! I am happy if I can find some small improvements in a PR

KGonzalezASC commented on Engine/Extensions/ExceptionExtensions.cs:
```diff
@@ -1,15 +1,20 @@
-namespace System
+namespace DotNetKoans.Engine.Extensions
 {
     public static class ExceptionExtensions
```
I haven't done unit tests before but I am willing to give it a try! Any suggestions or expectations for it?

MaxAtoms replied:
In that case a simple static method like this one is the perfect place to get started!
I guess one simple test case is sufficient.
You pass the static method a sample stack trace string and verify that the returned path strings are like expected.

Other tests are located in the directory `Engine/Tests`, by the way.
You should place your test there too.

KGonzalezASC committed:
-Included changes suggested by @MaxAtoms to remove capture group, and use quantifier +.
   -Additionally compiles Regex at runtime for efficency
-Included suggestions by @MaxAtoms to use records
   -Changed the LINQ expression to not reduce number of intermediate allocations
    otherPaths, and Where() filtered output.
   -Now iterates through allpaths once through aggregration

KGonzalezASC committed:
Merge branch 'NotMyself:master' into Feature/#186-only-relevant-stack-trace-paths

KGonzalezASC commented:
List of changes:
-Included changes suggested by @MaxAtoms to remove capture group, and use quantifier +.
   -Additionally compiles Regex at runtime for efficency
-Included suggestions by @MaxAtoms to use records
   -Changed the LINQ expression to not reduce number of intermediate allocations
    otherPaths, and Where() filtered output.
   -Now iterates through allpaths once through aggregation
   
Please let me know if there are any issues!

MaxAtoms reviewed (approved):
Thanks for the changes. They look good and correct :)

A couple of words about the optimizations:

- Here, `RegexOptions.Compiled` will probably make the code slower as we are using the regex only once, but the regex compilation takes more time. The fastest option would probably be a source generated regex (https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators), but it is really not needed for a simple one user CLI app like this.
- Second, I am not sure whether doing everything in one LINQ chain is faster here. In any case, it creates more objects because of the repeated usage of the `with` expression. Again, this is a sort of optimization that is not needed for a one user app and a list with only three elements. If you really needed to optimize performance in a hotpath of your app, I would consider dropping LINQ completely and fall back to manual iteration.

If you are interested in these sort of considerations, I highly recommend the videos by Nick Chapsas (e.g. https://youtu.be/jUZ3VKFyB-A). You could confirm whether my assumptions or yours are correct using Benchmark.NET.
